In [1]:
import pandas as pd

df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.shape

(7043, 21)

In [2]:
df['TotalCharges'] = df['TotalCharges'].replace(' ', '0')
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])
df['TotalCharges'].dtype

dtype('float64')

In [3]:
df[df['tenure'] == 0][['tenure', 'TotalCharges']]

,tenure,TotalCharges
488,0,0.0
753,0,0.0
936,0,0.0
1082,0,0.0
1340,0,0.0
3331,0,0.0
3826,0,0.0
4380,0,0.0
5218,0,0.0
6670,0,0.0


In [4]:
df = df.drop('customerID', axis=1)
df.shape

(7043, 20)

In [5]:
pd.get_dummies(df, columns=['InternetService'])

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,...,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,InternetService_DSL,InternetService_Fiber optic,InternetService_No
0,Female,0,Yes,No,1,No,No phone service,No,Yes,No,...,No,Month-to-month,Yes,Electronic check,29.85,29.85,No,True,False,False
1,Male,0,No,No,34,Yes,No,Yes,No,Yes,...,No,One year,No,Mailed check,56.95,1889.50,No,True,False,False
2,Male,0,No,No,2,Yes,No,Yes,Yes,No,...,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,True,False,False
3,Male,0,No,No,45,No,No phone service,Yes,No,Yes,...,No,One year,No,Bank transfer (automatic),42.30,1840.75,No,True,False,False
4,Female,0,No,No,2,Yes,No,No,No,No,...,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,Male,0,Yes,Yes,24,Yes,Yes,Yes,No,Yes,...,Yes,One year,Yes,Mailed check,84.80,1990.50,No,True,False,False
7039,Female,0,Yes,Yes,72,Yes,Yes,No,Yes,Yes,...,Yes,One year,Yes,Credit card (automatic),103.20,7362.90,No,False,True,False
7040,Female,0,Yes,Yes,11,No,No phone service,Yes,No,No,...,No,Month-to-month,Yes,Electronic check,29.60,346.45,No,True,False,False
7041,Male,1,Yes,No,4,Yes,Yes,No,No,No,...,No,Month-to-month,Yes,Mailed check,74.40,306.60,Yes,False,True,False


In [6]:
pd.get_dummies(df['InternetService'])

,DSL,Fiber optic,No
0,True,False,False
1,True,False,False
2,True,False,False
3,True,False,False
4,False,True,False
...,...,...,...
7038,True,False,False
7039,False,True,False
7040,True,False,False
7041,False,True,False


In [7]:
pd.get_dummies(df['InternetService'], drop_first=True)

,Fiber optic,No
0,False,False
1,False,False
2,False,False
3,False,False
4,True,False
...,...,...
7038,False,False
7039,True,False
7040,False,False
7041,True,False


In [8]:
df.dtypes

gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges        float64
Churn                   str
dtype: object

In [9]:
df['OnlineSecurity'].unique()

<StringArray>
['No', 'Yes', 'No internet service']
Length: 3, dtype: str

In [10]:
cols_to_fix = ['OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 
               'TechSupport', 'StreamingTV', 'StreamingMovies']

for col in cols_to_fix:
    df[col] = df[col].replace('No internet service', 'No')

df['MultipleLines'] = df['MultipleLines'].replace('No phone service', 'No')

In [11]:
df['OnlineSecurity'].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

In [12]:
df['PaymentMethod'].unique()

<StringArray>
[         'Electronic check',              'Mailed check',
 'Bank transfer (automatic)',   'Credit card (automatic)']
Length: 4, dtype: str

In [13]:
binary_cols = ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
               'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
               'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'Churn']

for col in binary_cols:
    df[col] = df[col].map({'Yes': 1, 'No': 0, 'Male': 1, 'Female': 0})

In [14]:
df[binary_cols].head()

,gender,Partner,Dependents,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,Churn
0,0,1,0,0,0,0,1,0,0,0,0,1,0
1,1,0,0,1,0,1,0,1,0,0,0,0,0
2,1,0,0,1,0,1,1,0,0,0,0,1,1
3,1,0,0,0,0,1,0,1,1,0,0,0,0
4,0,0,0,1,0,0,0,0,0,0,0,1,1


In [15]:
df['PaymentMethod'].unique()

<StringArray>
[         'Electronic check',              'Mailed check',
 'Bank transfer (automatic)',   'Credit card (automatic)']
Length: 4, dtype: str

In [16]:
df = pd.get_dummies(df, columns=['InternetService', 'Contract', 'PaymentMethod'], drop_first=True)

In [17]:
df.shape

(7043, 24)

In [18]:
df.columns.tolist()

['gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'PaperlessBilling',
 'MonthlyCharges',
 'TotalCharges',
 'Churn',
 'InternetService_Fiber optic',
 'InternetService_No',
 'Contract_One year',
 'Contract_Two year',
 'PaymentMethod_Credit card (automatic)',
 'PaymentMethod_Electronic check',
 'PaymentMethod_Mailed check']

In [19]:
df.dtypes

gender                                     int64
SeniorCitizen                              int64
Partner                                    int64
Dependents                                 int64
tenure                                     int64
PhoneService                               int64
MultipleLines                              int64
OnlineSecurity                             int64
OnlineBackup                               int64
DeviceProtection                           int64
TechSupport                                int64
StreamingTV                                int64
StreamingMovies                            int64
PaperlessBilling                           int64
MonthlyCharges                           float64
TotalCharges                             float64
Churn                                      int64
InternetService_Fiber optic                 bool
InternetService_No                          bool
Contract_One year                           bool
Contract_Two year   

In [20]:
df.to_csv('../data/processed/telco_churn_cleaned.csv', index=False)

## Data Cleaning Summary

Cleaned the raw Telco churn dataset in preparation for modeling:

- **Fixed `TotalCharges`**: 11 rows had blank values due to `tenure = 0` (brand new customers 
  with no billing history yet). Replaced with `0` and converted the column from text to numeric, 
  rather than dropping these rows. Model should handle bran new customers correctly. 
- **Dropped `customerID`**: a unique, non predictive identifier that risked encouraging the model 
  to memorize individual rows rather than learn generalizable patterns.
- **Simplified redundant categories**: collapsed "No internet service" / "No phone service" labels 
  down to plain "No" across 7 columns, since that information was already fully captured by 
  `InternetService`/`PhoneService`.
- **Encoded categorical variables**: binary Yes/No (and gender) columns mapped to 0/1; multi category 
  columns (`InternetService`, `Contract`, `PaymentMethod`) one hot encoded with `drop_first=True` 
  to avoid redundant, perfectly correlated columns.

Final dataset: 7043 rows × 24 columns, fully numeric and ready for train/test splitting.